fuse_constraints/src/relative_orientation_3d_stamped_constraint.cpp

```diff
@@ -0,0 +1,118 @@
+#include <Eigen/Geometry>
+
+
+namespace fuse_constraints
+{
+
+RelativeOrientation3DStampedConstraint::RelativeOrientation3DStampedConstraint(
+  const fuse_variables::Orientation3DStamped& orientation1,
+  const fuse_variables::Orientation3DStamped& orientation2,
+  const fuse_core::Vector4d& delta,
+  const fuse_core::Matrix3d& covariance) :
+    fuse_core::Constraint{orientation1.uuid(), orientation2.uuid()},
+    delta_(delta),
+    sqrt_information_(covariance.inverse().llt().matrixU())
+{
+}
+
+RelativeOrientation3DStampedConstraint::RelativeOrientation3DStampedConstraint(
+  const fuse_variables::Orientation3DStamped& orientation1,
+  const fuse_variables::Orientation3DStamped& orientation2,
+  const Eigen::Quaterniond& delta,
+  const fuse_core::Matrix3d& covariance) :
+    RelativeOrientation3DStampedConstraint(orientation1, orientation2, toEigen(delta), covariance)
+{
+}
+
+RelativeOrientation3DStampedConstraint::RelativeOrientation3DStampedConstraint(
+  const fuse_variables::Orientation3DStamped& orientation1,
+  const fuse_variables::Orientation3DStamped& orientation2,
+  const geometry_msgs::Quaternion& delta,
+  const std::array<double, 9>& covariance) :
+    RelativeOrientation3DStampedConstraint(orientation1, orientation2, toEigen(delta), toEigen(covariance))
+{
+}
+
+fuse_core::Matrix3d RelativeOrientation3DStampedConstraint::covariance() const
+{
+  return (sqrt_information_.transpose() * sqrt_information_).inverse();
+}
+
+void RelativeOrientation3DStampedConstraint::print(std::ostream& stream) const
+{
+  stream << type() << "\n"
+         << "  uuid: " << uuid() << "\n"
+         << "  orientation variable1: " << variables_.at(0) << "\n"
+         << "  orientation variable2: " << variables_.at(1) << "\n"
+         << "  delta: " << delta_.transpose() << "\n"
+         << "  sqrt_info: " << sqrtInformation() << "\n";
+}
+
+fuse_core::Constraint::UniquePtr RelativeOrientation3DStampedConstraint::clone() const
+{
+  return RelativeOrientation3DStampedConstraint::make_unique(*this);
+}
+
+ceres::CostFunction* RelativeOrientation3DStampedConstraint::costFunction() const
+{
+  return new ceres::AutoDiffCostFunction<NormalDeltaOrientation3DCostFunctor, 3, 4, 4>(
+    new NormalDeltaOrientation3DCostFunctor(sqrt_information_, delta_));
+}
+
+fuse_core::Vector4d RelativeOrientation3DStampedConstraint::toEigen(const Eigen::Quaterniond& quaternion)
+{
+  fuse_core::Vector4d eigen_quaternion_vector;
+  eigen_quaternion_vector << quaternion.w(), quaternion.x(), quaternion.y(), quaternion.z();
+  return eigen_quaternion_vector;
+}
+
+fuse_core::Vector4d RelativeOrientation3DStampedConstraint::toEigen(const geometry_msgs::Quaternion& quaternion)
+{
+  fuse_core::Vector4d eigen_quaternion_vector;
+  eigen_quaternion_vector << quaternion.w, quaternion.x, quaternion.y, quaternion.z;
+  return eigen_quaternion_vector;
+}
+
+fuse_core::Matrix3d RelativeOrientation3DStampedConstraint::toEigen(const std::array<double, 9>& covariance)
```

Comment:
Could this return a matrix that is an `Eigen::Map`? Would that be valuable at all?

Reply:
I don't think so...at least not safely. The input _could_ be a temporary.

Based on the signature, something like this should be legal:
```
auto cov_matrix = toEigen({1, 0, 0,  0, 1, 0,  0, 0, 1});
```

 Using Eigen::Map would result in pointing at data that was deleted once the function returns.